### PR TITLE
Nested directories dataset capability

### LINKED ISSUE #1379: Support nested directories in custom dataset training
## 🚀 Feature
Support nested folder structure to create dataset while training or testing model. Currently, dataset.py is finding files only at the top layer of the folder.

## Motivation
Data are still separated in train/test folder. However, adding a feature to [datasets.py](https://github.com/ultralytics/yolov5/blob/master/utils/datasets.py) to walk in nested folder structure will help us to train network without much modification on the dataset directories. 

I'm getting error because of the following line,
https://github.com/ultralytics/yolov5/blob/04081f810270712ba3a69577c47e5dcfa850fa90/utils/datasets.py#L357

eg. My image folder looks like
```
train
|___cat
    |____image1.png
    |____image2.png
    |____image3.png
|___dog
```
and the same structure in the test folder.

### PR DESCRIPTION
**Pull request by glenn-jocher**
This PR adds experimental support for nested directory structure datasets. See https://github.com/ultralytics/yolov5/issues/1379

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Improved checkpoint and anchor configurations, data loading enhancements, and code modularization.

### 📊 Key Changes
- `.gitignore` updated to include `wandb/`
- Default anchor count in `hyp.scratch.yaml` changed from 0 to 3
- `detect.py` and other script imports and references adjusted for modularity
- YOLO `models/yolo.py` anchor checks and FLOPS calculation updated
- `test.py` updated to reflect changes in data loading and anchor handling, COCO dataset checking, and better Weights & Biases integration
- `train.py` refactored to consolidate loss computation, introduce `autoanchor` check, and clean up Weights & Biases handling
- Jupyter notebook `tutorial.ipynb` paths updated to new modules
- `activations.py`, `autoanchor.py` and `loss.py` files created for better code organization
- `datasets.py` changes reflect data loading and augmentation updates
- Shell script `evolve.sh` and a few utility functions removed as part of cleanup
- `general.py` utility functions adjusted and some moved to new module files for better structure
- `google_utils.py` dependencies comment corrected
- `metrics.py` module created for validation metrics functions
- `plots.py` module created to handle different types of plots

### 🎯 Purpose & Impact
- Enhances the modularity and maintainability of the codebase 🛠️
- Simplifies workflows like training and testing by improving utility functions 🚀
- Better integrates popular experiment tracking tool Weights & Biases for cooperative work and tracking 📊
- Updates default configuration and behavior for improved object detection training 🎯

### CODE REVIEW

File: utils/datasets.py [hunk 1]
```diff
@@ -347,19 +347,19 @@ def img2label_paths(img_paths):
...
+                p = Path(p)  # os-agnostic
+                if p.is_file():  # file
+                    parent = str(p.parent) + os.sep
                     with open(p, 'r') as t:
                         t = t.read().splitlines()
                         f += [x.replace('./', parent) if x.startswith('./') else x for x in t]  # local to global path
-                elif os.path.isdir(p):  # folder
-                    f += glob.iglob(p + os.sep + '*.*')
+                elif p.is_dir():  # dir
+                    f += glob.glob(str(p) + '/**/*.*', recursive=True)
```
Comment (priteshgohil) [inline utils/datasets.py]: Something is wrong at this line. Program hangs on here.

Comment (glenn-jocher) [reply utils/datasets.py]: What OS are you on? Perhaps I need to make this OS robust. I will push an update here hold on.

Comment (glenn-jocher) [reply utils/datasets.py]: Ok I've pushed ec75dde which may help on Windows machines. If you are still having problems please try to reproduce with COCO128 (i.e. by copying one COCO128 into another) on a Google Colab notebook and share.

File: utils/datasets.py [hunk 2]
```diff
@@ -347,19 +347,19 @@ def img2label_paths(img_paths):
...
+                p = Path(p)  # os-agnostic
+                if p.is_file():  # file
+                    parent = str(p.parent) + os.sep
                     with open(p, 'r') as t:
                         t = t.read().splitlines()
                         f += [x.replace('./', parent) if x.startswith('./') else x for x in t]  # local to global path
-                elif os.path.isdir(p):  # folder
-                    f += glob.iglob(p + os.sep + '*.*')
+                elif p.is_dir():  # dir
+                    f += glob.glob(f"{p}{os.sep}**{os.sep}*.*", recursive=True)
```
Comment (glenn-jocher) [inline utils/datasets.py:352]: Updated section of code is here.

Comment (glenn-jocher) [reply utils/datasets.py:352]: @priteshgohil can you check to see whether this PR is still giving you errors? I need to either merge or close this soon for housekeeping. Thanks.

**Pull request closed without merging**